### DISCUSSION TITLE
How to identify and remove dead end pores and isolated clusters after using reduce_coordination()

### DISCUSSION THREAD

**Discussion by harshpallam** [PMEAL/OpenPNM#2567] [Q&A] [answered]
I wanted to generate a random network with a mean coordination number of 3.9.
The initial size of the cubic network is 51 * 20 * 20 nodes in x,y, and z directions with a coordination number of 26 for all pore bodies.
I used `reduce_coordination()` to reduce the mean connectivity to 3.9. However, after coordination number reduction, I see that 20400 pores are still present and none of the pores are eliminated. When I tried to trim the pores which are not connected to the network and used `check_network_health()`, it showed an empty set. Please see the attached image. Does it mean that there are no isolated nodes or clusters present in the network?
I request you to kindly help me with this.
![Screenshot from 2022-09-16 17-01-06](https://user-images.githubusercontent.com/95679531/190705021-05c3cf33-bfa6-4fe2-ba70-f834d9b0f61f.png)
```
import os
import warnings
import numpy as np
import openpnm as op
import scipy as sp
import random
import matplotlib.pyplot as plt
import openpnm.models as mods
import sys
# import openpnm.physics.models as pm
np.random.seed(11)
ws = op.Workspace()
ws.settings['loglevel'] = 40
warnings.filterwarnings('ignore')
#######################################################################
#1 setting up network
pn = op.network.Cubic(shape=[50, 20, 20], spacing=[0.001, 0.001, 0.001], connectivity=26) #add_boundary_pores=['left', 'right']
op.io.VTK.save(pn, filename="hp")
op.io.CSV.save(pn, filename="hp")
print(pn.am)
print(pn.conns)
print(pn.coords)
print(pn.im)
print("No of throats before elem =",pn.num_throats())
print("No of pores before elem =",pn.num_pores())
print(f"Does 'hp.vtk' exist? {os.path.isfile('hp.vtp')}")
print("A Cubic Lattice is created!")
#Elimination of network
op.topotools.reduce_coordination(network=pn, z=3.9)
print(pn['throat.conns'])
#checking the pore connectivity
h = pn.check_network_health()
op.topotools.trim(network=pn, pores=h['trim_pores'])
h = pn.check_network_health()
print(h)
op.io.VTK.save(pn, filename="hp_r")
op.io.CSV.save(pn, filename="hp_r")
print("No of throats after elem =",pn.num_throats())
print("No of pores after elem =",pn.num_pores())
```

**Comment by jgostick** [comment, marked as answer]
Isolate clusters is easy...these should show up on the ``check_network_health`` function.  Deadend pores are not technically a problem for the simulations, so we consider those as healthy.  However, if you really don't want deadend pores you can do ``Ps = pn.num_neighbors(pores=np.Ps)`` which will return a list of the number of neighbors to each pore.  You can then trim any pores with only 1 neighbor (``op.topotools.trim(pn, pores=Ps==1)``.  You'll need to do this several times since you may create new deadend pores upon deletion of the previous ones (as would happen in the case of a chain of pores leading to a dead end).

**Comment by jgostick** [comment]
Oh, and I should also mention that trim coordination does not remove any pores, only throats, AND it does it in a way that maintains connectivity.  It finds the "minimum spanning tree" of the network (assuming random throat weights), then only deletes throats that are NOT part of the spanning tree.

**Comment by harshpallam** [comment]
Thanks a lot, sir. It worked. `Ps = pn.num_neighbors(pores=pn.Ps)`
